### PR TITLE
Create an Engine Connector

### LINKED ISSUE #192: [v0.3.0] Create an PluginConnector
implements #183

Create a `PluginConnector` for the engine side : 
```typescript
interface PluginConnector {
  /** Send a message to the external plugin */
  send(message: Partial<Message>): void
  /** Get messae from the external plugin */
  on((message: Partial<Message>) => void): void
  /** Verify if the plugin is currently connected */
  isConnected(): boolean
  /** Open connection with the plugin */
  connect()
  /** Close connection with the plugin */
  disconnect()
}
```

Implement it in : 
- [ ] Iframe
- [ ] Websocket
- [ ] Child process

### PR DESCRIPTION
**Pull request by GrandSchtroumpf**
Create 
- [X] `PluginConnector` in engine
- [X] `@remixproject/engine-web`  with Websocket & Iframe
- [X] `@remixproject/engine-node` with Child Process

fix #192

### CODE REVIEW

**Comment by GrandSchtroumpf** [comment]
This new implementation create a lot of build issues : 
- We don't want to build the engine with each package.
- The engine & plugins rely on the "utils" library which is not an npm package.
-> Each `@remixproject/engine-*` will include duplicate code from the utils package.

One solution would be to create a package `@remixproject/plugin-utils`.  But then each project would depends on `@remixproject/plugin-utils` even if they don't need all of it...

**Comment by GrandSchtroumpf** [comment]
It seems that it's not a real problem yet as the only import required from the utils library are types, so they have no impact on the bundle size.
Also using [externals from webpack](https://webpack.js.org/configuration/externals/) removed the code from `@remixproject/engine`.

**Comment by yann300** [comment]
It looks like this PR encapsulate each plugin connection into a `PluginConnector` but doesn't change the logic, am I right?
also added @ioedeveloper @Aniket-Engg to the PR

**Comment by GrandSchtroumpf** [comment]
Yes, this is the idea. It's "just" a way to extract the plugin connection logic outside of the core engine. Like that Cale from microsoft will be able to create his own set of plugin connector

File: projects/engine/src/connector.ts [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+  }
+
+  deactivate() {
+    this.loaded = false
+    this.disconnect()
+    super.deactivate()
+  }
+
+  /** Call a method from this plugin */
+  protected callPluginMethod(key: string, payload: any[] = []): Promise<any> {
```
[the remark this answers is not in the document]
Comment (GrandSchtroumpf) [reply projects/engine/src/connector.ts:53]: I don't really see how that could be possible. 
The payload **can** be anything (null, undefined, function, array, object, string, number, ...) & the result value also. This is actually the definition of `any`.

[the remark this answers is not in the document]
Comment (ioedeveloper) [reply projects/engine/src/connector.ts:53]: Okay, then i guess the same applies to the return type.

**Pull request merged**